### ISSUE TITLE
Test suite failure on GHC 7.6

### ISSUE THREAD

**Issue by snoyberg** [hspec/hspec#210]
I seem to remember mentioning this somewhere else, so if I have, apologies for the dupe. I just got this failure from the Stackage GHC 7.6 run:

```
Resolving dependencies...
Configuring hspec-core-2.0.2...
Running Haddock for hspec-core-2.0.2...
Running hscolour for hspec-core-2.0.2...
Preprocessing library hspec-core-2.0.2...
Preprocessing test suite 'spec' for hspec-core-2.0.2...
Preprocessing library hspec-core-2.0.2...
Warning: The documentation for the following packages are not installed. No
links will be generated to these packages: HUnit-1.2.5.2, QuickCheck-2.7.6,
ansi-terminal-0.6.2.1, async-2.0.1.6, hspec-expectations-0.6.1,
primitive-0.5.4.0, quickcheck-io-0.1.1, random-1.0.1.1, setenv-0.1.1.1,
stm-2.4.3, tf-random-0.5, transformers-0.3.0.0
Haddock coverage:
   0% (  0 /  2) in 'Test.Hspec.Timer'
  50% (  3 /  6) in 'Test.Hspec.Compat'
 100% ( 12 / 12) in 'Test.Hspec.Core.Util'
   0% (  0 /  7) in 'Test.Hspec.Core.QuickCheckUtil'
   0% (  0 /  4) in 'Test.Hspec.FailureReport'
Warning: Test.Hspec.Core.Example: We do not support associated types in instances yet. These instances are affected:
Test.Hspec.Core.Example.Example GHC.Types.Bool, Test.Hspec.Core.Example.Example Test.Hspec.Expectations.Expectation, Test.Hspec.Core.Example.Example (a
                                 -> Test.Hspec.Expectations.Expectation), Test.Hspec.Core.Example.Example Test.Hspec.Core.Example.Result, Test.Hspec.Core.Example.Example Test.QuickCheck.Property.Property, Test.Hspec.Core.Example.Example (a
                                 -> Test.QuickCheck.Property.Property)
  38% (  3 /  8) in 'Test.Hspec.Core.Example'
 100% (  8 /  8) in 'Test.Hspec.Core.Tree'
  36% (  4 / 11) in 'Test.Hspec.Core.Spec.Monad'
  71% ( 24 / 34) in 'Test.Hspec.Core.Spec'
 100% ( 11 / 11) in 'Test.Hspec.Core.Hooks'
doc comment parse failed: use @writeLine ""@ instead
  81% ( 22 / 27) in 'Test.Hspec.Core.Formatters.Internal'
  77% ( 24 / 31) in 'Test.Hspec.Core.Formatters'
   0% (  0 /  6) in 'Test.Hspec.Options'
  14% (  1 /  7) in 'Test.Hspec.Config'
  50% (  1 /  2) in 'Test.Hspec.Core.Runner.Eval'
  77% ( 10 / 13) in 'Test.Hspec.Core.Runner'
 100% (  4 /  4) in 'Test.Hspec.Core.QuickCheck'
Documentation created: dist/doc/html/hspec-core/index.html,
dist/doc/html/hspec-core/hspec-core.txt
Preprocessing test suite 'spec' for hspec-core-2.0.2...
Building hspec-core-2.0.2...
Preprocessing library hspec-core-2.0.2...
[ 1 of 17] Compiling Test.Hspec.Timer ( src/Test/Hspec/Timer.hs, dist/build/Test/Hspec/Timer.o )
[ 2 of 17] Compiling Test.Hspec.Compat ( src/Test/Hspec/Compat.hs, dist/build/Test/Hspec/Compat.o )
[ 3 of 17] Compiling Test.Hspec.Core.Util ( src/Test/Hspec/Core/Util.hs, dist/build/Test/Hspec/Core/Util.o )
[ 4 of 17] Compiling Test.Hspec.Core.QuickCheckUtil ( src/Test/Hspec/Core/QuickCheckUtil.hs, dist/build/Test/Hspec/Core/QuickCheckUtil.o )
[ 5 of 17] Compiling Test.Hspec.FailureReport ( src/Test/Hspec/FailureReport.hs, dist/build/Test/Hspec/FailureReport.o )
[ 6 of 17] Compiling Test.Hspec.Core.Example ( src/Test/Hspec/Core/Example.hs, dist/build/Test/Hspec/Core/Example.o )
[ 7 of 17] Compiling Test.Hspec.Core.Tree ( src/Test/Hspec/Core/Tree.hs, dist/build/Test/Hspec/Core/Tree.o )
[ 8 of 17] Compiling Test.Hspec.Core.Spec.Monad ( src/Test/Hspec/Core/Spec/Monad.hs, dist/build/Test/Hspec/Core/Spec/Monad.o )
[ 9 of 17] Compiling Test.Hspec.Core.Spec ( src/Test/Hspec/Core/Spec.hs, dist/build/Test/Hspec/Core/Spec.o )
[10 of 17] Compiling Test.Hspec.Core.Hooks ( src/Test/Hspec/Core/Hooks.hs, dist/build/Test/Hspec/Core/Hooks.o )
[11 of 17] Compiling Test.Hspec.Core.Formatters.Internal ( src/Test/Hspec/Core/Formatters/Internal.hs, dist/build/Test/Hspec/Core/Formatters/Internal.o )
[12 of 17] Compiling Test.Hspec.Core.Formatters ( src/Test/Hspec/Core/Formatters.hs, dist/build/Test/Hspec/Core/Formatters.o )
[13 of 17] Compiling Test.Hspec.Options ( src/Test/Hspec/Options.hs, dist/build/Test/Hspec/Options.o )
[14 of 17] Compiling Test.Hspec.Config ( src/Test/Hspec/Config.hs, dist/build/Test/Hspec/Config.o )
[15 of 17] Compiling Test.Hspec.Core.Runner.Eval ( src/Test/Hspec/Core/Runner/Eval.hs, dist/build/Test/Hspec/Core/Runner/Eval.o )
[16 of 17] Compiling Test.Hspec.Core.Runner ( src/Test/Hspec/Core/Runner.hs, dist/build/Test/Hspec/Core/Runner.o )
[17 of 17] Compiling Test.Hspec.Core.QuickCheck ( src/Test/Hspec/Core/QuickCheck.hs, dist/build/Test/Hspec/Core/QuickCheck.o )
In-place registering hspec-core-2.0.2...
Preprocessing test suite 'spec' for hspec-core-2.0.2...
[ 1 of 31] Compiling Mock             ( test/Mock.hs, dist/build/spec/spec-tmp/Mock.o )
[ 2 of 31] Compiling Test.Hspec.Timer ( src/Test/Hspec/Timer.hs, dist/build/spec/spec-tmp/Test/Hspec/Timer.o )
[ 3 of 31] Compiling Test.Hspec.Compat ( src/Test/Hspec/Compat.hs, dist/build/spec/spec-tmp/Test/Hspec/Compat.o )
[ 4 of 31] Compiling Test.Hspec.Core.Util ( src/Test/Hspec/Core/Util.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/Util.o )
[ 5 of 31] Compiling Test.Hspec.Core.QuickCheckUtil ( src/Test/Hspec/Core/QuickCheckUtil.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/QuickCheckUtil.o )
[ 6 of 31] Compiling Test.Hspec.Core.Example ( src/Test/Hspec/Core/Example.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/Example.o )
[ 7 of 31] Compiling Test.Hspec.Core.Tree ( src/Test/Hspec/Core/Tree.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/Tree.o )
[ 8 of 31] Compiling Test.Hspec.Core.Spec.Monad ( src/Test/Hspec/Core/Spec/Monad.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/Spec/Monad.o )
[ 9 of 31] Compiling Test.Hspec.FailureReport ( src/Test/Hspec/FailureReport.hs, dist/build/spec/spec-tmp/Test/Hspec/FailureReport.o )
[10 of 31] Compiling Test.Hspec.Core.Spec ( src/Test/Hspec/Core/Spec.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/Spec.o )
[11 of 31] Compiling Test.Hspec.Core.Hooks ( src/Test/Hspec/Core/Hooks.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/Hooks.o )
[12 of 31] Compiling Test.Hspec.Core.QuickCheck ( src/Test/Hspec/Core/QuickCheck.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/QuickCheck.o )
[13 of 31] Compiling Test.Hspec.Core.Formatters.Internal ( src/Test/Hspec/Core/Formatters/Internal.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/Formatters/Internal.o )
[14 of 31] Compiling Test.Hspec.Core.Formatters ( src/Test/Hspec/Core/Formatters.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/Formatters.o )
[15 of 31] Compiling Test.Hspec.Options ( src/Test/Hspec/Options.hs, dist/build/spec/spec-tmp/Test/Hspec/Options.o )
[16 of 31] Compiling Test.Hspec.Config ( src/Test/Hspec/Config.hs, dist/build/spec/spec-tmp/Test/Hspec/Config.o )
[17 of 31] Compiling Test.Hspec.Core.Runner.Eval ( src/Test/Hspec/Core/Runner/Eval.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/Runner/Eval.o )
[18 of 31] Compiling Test.Hspec.Core.Runner ( src/Test/Hspec/Core/Runner.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/Runner.o )
[19 of 31] Compiling Helper           ( test/Helper.hs, dist/build/spec/spec-tmp/Helper.o )
[20 of 31] Compiling Test.Hspec.TimerSpec ( test/Test/Hspec/TimerSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/TimerSpec.o )
[21 of 31] Compiling Test.Hspec.OptionsSpec ( test/Test/Hspec/OptionsSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/OptionsSpec.o )
[22 of 31] Compiling Test.Hspec.FailureReportSpec ( test/Test/Hspec/FailureReportSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/FailureReportSpec.o )
[23 of 31] Compiling Test.Hspec.Core.UtilSpec ( test/Test/Hspec/Core/UtilSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/UtilSpec.o )
[24 of 31] Compiling Test.Hspec.Core.SpecSpec ( test/Test/Hspec/Core/SpecSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/SpecSpec.o )
[25 of 31] Compiling Test.Hspec.Core.RunnerSpec ( test/Test/Hspec/Core/RunnerSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/RunnerSpec.o )
[26 of 31] Compiling Test.Hspec.Core.QuickCheckUtilSpec ( test/Test/Hspec/Core/QuickCheckUtilSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/QuickCheckUtilSpec.o )
[27 of 31] Compiling Test.Hspec.Core.HooksSpec ( test/Test/Hspec/Core/HooksSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/HooksSpec.o )
[28 of 31] Compiling Test.Hspec.Core.FormattersSpec ( test/Test/Hspec/Core/FormattersSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/FormattersSpec.o )
[29 of 31] Compiling Test.Hspec.Core.ExampleSpec ( test/Test/Hspec/Core/ExampleSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/Core/ExampleSpec.o )
[30 of 31] Compiling Test.Hspec.CompatSpec ( test/Test/Hspec/CompatSpec.hs, dist/build/spec/spec-tmp/Test/Hspec/CompatSpec.o )
[31 of 31] Compiling Main             ( test/Spec.hs, dist/build/spec/spec-tmp/Main.o )
Linking dist/build/spec/spec ...
Building hspec-core-2.0.2...
Preprocessing library hspec-core-2.0.2...
In-place registering hspec-core-2.0.2...
Preprocessing test suite 'spec' for hspec-core-2.0.2...
Running 1 test suites...
Test suite spec: RUNNING...

Test.Hspec.Compat
  showType
    shows unqualified name of type
  showFullType (currently unused)
    shows fully qualified name of type
  lookupEnv
    returns value of specified environment variable
    returns Nothing if specified environment variable is not set
Test.Hspec.Core.Example
  evaluateExample
    for Bool
      returns Success on True
      returns Fail on False
      propagates exceptions
    for Expectation
      returns Success if all expectations hold
      returns Fail if an expectation does not hold
      propagates exceptions
      runs provided action around expectation
      when used with `pending`
        returns Pending
      when used with `pendingWith`
        includes the optional reason
    for Property
      returns Success if property holds
      returns Fail if property does not hold
      shows what falsified it
      runs provided action around each single check of the property
      pretty-prints exceptions
      when used with shouldBe
        shows what falsified it
      when used with `pending`
        returns Pending
      when used with `pendingWith`
        includes the optional reason
  Expectation
    as a QuickCheck property
      can be quantified
      can be used with expectations/HUnit assertions
Test.Hspec.Core.Formatters
  silent
    produces no output
  failed_examples
    summarizes the time it takes to finish
    displays a detailed list of failures
      prints all requirements that are not met
      prints the exception type for requirements that fail due to an uncaught exception
      prints all descriptions when a nested requirement fails
      when a failed example has a source location
        when source location is exact
          includes that source locations
          does not include 'best-effort' explanation
        when source location is best-effort
          marks that source location as 'best-effort'
          includes 'best-effort' explanation
    summarizes the number of examples and failures
    shows summary in green if there are no failures
    shows summary in yellow if there are pending examples
    shows summary in red if there are failures
    shows summary in red if there are both failures and pending examples
  progress
    produces '..F...FF.F' style output
    same as failed_examples
      summarizes the time it takes to finish
      displays a detailed list of failures
        prints all requirements that are not met
        prints the exception type for requirements that fail due to an uncaught exception
        prints all descriptions when a nested requirement fails
        when a failed example has a source location
          when source location is exact
            includes that source locations
            does not include 'best-effort' explanation
          when source location is best-effort
            marks that source location as 'best-effort'
            includes 'best-effort' explanation
      summarizes the number of examples and failures
      shows summary in green if there are no failures
      shows summary in yellow if there are pending examples
      shows summary in red if there are failures
      shows summary in red if there are both failures and pending examples
  specdoc
    displays a header for each thing being described
    displays one row for each behavior
    outputs an empty line at the beginning (even for non-nested specs)
    displays a row for each successfull, failed, or pending example
    displays a '#' with an additional message for pending examples
    with an empty group
      omits that group from the report
    same as failed_examples
      summarizes the time it takes to finish
      displays a detailed list of failures
        prints all requirements that are not met
        prints the exception type for requirements that fail due to an uncaught exception
        prints all descriptions when a nested requirement fails
        when a failed example has a source location
          when source location is exact
            includes that source locations
            does not include 'best-effort' explanation
          when source location is best-effort
            marks that source location as 'best-effort'
            includes 'best-effort' explanation
      summarizes the number of examples and failures
      shows summary in green if there are no failures
      shows summary in yellow if there are pending examples
      shows summary in red if there are failures
      shows summary in red if there are both failures and pending examples
Test.Hspec.Core.Hooks
  before
    runs an action before every spec item
    when used multiple times
      is evaluated outside in
    when used with an action that returns a value
      passes that value to the spec item
    when used with a QuickCheck property
      runs action before every check of the property
  beforeWith
    runs an action before every spec item
  beforeAll
    runs an action before the first spec item
    when used with an empty list of examples
      does not run specified action
    when used with an action that returns a value
      passes that value to the spec item
  after
    must be used with a before action
  after_
    runs an action after every spec item
    guarantees that action is run
    when used multiple times
      is evaluated inside out
  afterAll_
    runs an action after the last spec item
    when used with an empty list of examples
      does not run specified action
    when action throws an exception
      reports a failure
  around
    wraps every spec item with an action
  around_
    wraps every spec item with an action
    when used multiple times
      is evaluated outside in
  aroundWith
    wraps every spec item with an action
Test.Hspec.Core.QuickCheckUtil
  formatNumbers
    includes number of tests
    pluralizes number of tests
    includes number of shrinks
    pluralizes number of shrinks
Test.Hspec.Core.Runner
  hspec
    runs a spec
    exits with exitFailure if not all examples pass
    allows output to stdout
    prints an error message on unrecognized command-line options
    stores a failure report in the environment
    with --rerun
      reruns examples that previously failed
      reuses the same seed
      reuses same --qc-max-success
      reuses same --qc-max-size
      reuses same --qc-max-discard
      when there is no failure report in the environment
        runs everything
        prints a warning to stderr
      when parsing of failure report fails
        runs everything
        prints a warning to stderr
    does not leak command-line flags to examples
    when interrupted with ctrl-c
      prints summary immediately
      throws UserInterrupt
    with --help
      prints help
      constrains lines to 80 characters
    with --dry-run
      produces a report
      does not verify anything
      ignores afterAll-hooks
    with --fail-fast
      stops after first failure
      works for nested specs
    with --match
      only runs examples that match a given pattern
      can be given multiple times
    with --format
      uses specified formatter
      when given an invalid argument
        prints an error message to stderr
    with --qc-max-success
      tries QuickCheck properties specified number of times
      when run with --rerun
        takes precedence
    with --qc-max-size
      passes specified size to QuickCheck properties
    with --qc-max-discard
      uses specified discard ratio to QuickCheck properties
    with --seed
      uses specified seed
      when run with --rerun
        takes precedence
      when given an invalid argument
        prints an error message to stderr
        exits with exitFailure
    with --print-cpu-time
      includes used CPU time in summary
    with --html
      produces HTML output
      marks successful examples with CSS class hspec-success
      marks pending examples with CSS class hspec-pending
      marks failed examples with CSS class hspec-failure
  hspecResult
    returns a summary of the test run
    treats uncaught exceptions as failure
    uses the specdoc formatter by default
    can use a custom formatter
    does not let escape error thunks from failure messages
    runs specs in parallel
Test.Hspec.Core.Spec
  describe
    can be nested
    when no description is given
      uses a default description
  it
    takes a description of a desired behavior
    takes an example of that behavior
    when no description is given
      uses a default description
  pending
    specifies a pending example
  pendingWith

<no location info>:
    Could not find module `System.SetEnv'
    It is not a module in the current program, or in any known package.

<interactive>:3:1: Not in scope: `setEnv'
*** Exception: FOO: getEnv: does not exist (no environment variable)
    specifies a pending example with a reason for why it's pending
  parallel
    marks examples for parallel execution
    is applied recursively
Test.Hspec.Core.Util
  pluralize
    returns singular when used with 1
    returns plural when used with number greater 1
    returns plural when used with 0
  formatException
    converts exception to string
    when used with an IOException
      includes the IOErrorType
  lineBreaksAt
    inserts line breaks at word boundaries
  safeTry
    returns Right on success
    returns Left on exception
    evaluates result to weak head normal form
    does not catch asynchronous exceptions
  filterPredicate
    tries to match a pattern against a path
    is ambiguous
    succeeds on a partial match
    succeeds with a pattern that matches the message give in the failure list
  formatRequirement
    creates a sentence from a subject and a requirement
    creates a sentence from a subject and a requirement when the subject consits of multiple words
    returns the requirement if no subject is given
    inserts context separated by commas
    joins components of a subject with a dot
    properly handles context after a subject that consists of several components
Test.Hspec.FailureReport
  writeFailureReport
    prints a warning on unexpected exceptions
  GHCi
    keeps environment variables on :reload FAILED [1]
Test.Hspec.Options
  parseOptions
    sets configColorMode to ColorAuto
    with --no-color
      sets configColorMode to ColorNever
    with --color
      sets configColorMode to ColorAlways
    with --out
      sets configOutputFile
    with --qc-max-success
      when given an invalid argument
        returns an error message
    with --depth
      sets depth parameter for SmallCheck
Test.Hspec.Timer
  timer action returned by newTimer
    returns False
    after specified time
      returns True

1) Test.Hspec.FailureReport.GHCi keeps environment variables on :reload
expected: "\"bar\"\n"
 but got: ""

# test/Test/Hspec/FailureReportSpec.hs:25 (best-effort)

Source locations marked with "best-effort" are calculated heuristically and may be incorrect.

Randomized with seed 447258390

Finished in 1.3136 seconds
179 examples, 1 failure
Test suite spec: FAIL
Test suite logged to: dist/test/hspec-core-2.0.2-spec.log
0 of 1 test suites (0 of 1 test cases) passed.
```
